internal/app/go-filecoin/retrieval_market_connector/client.go

```diff
@@ -1,27 +1,292 @@
+	lane := uint64(len(r.laneStore[payer]) - 1)
+
+	return lane, nil
+}
+
+// CreatePaymentVoucher tells the actor to create and sign a voucher and post this to the chain, returning
+// the signed voucher and any error.
+// If there is not enough value stored in the payment channel registry, an error is returned.
+// If a lane has not been allocated for this payment channel, an error is returned.
+func (r *RetrievalClientNodeConnector) CreatePaymentVoucher(ctx context.Context, paymentChannel address.Address, amount tokenamount.TokenAmount, lane uint64) (*gfm_types.SignedVoucher, error) {
+	payer, err := r.getPayerFromChannelID(paymentChannel)
+	if err != nil {
+		return nil, err
+	}
+
+	_, pinfo, err := r.pbAPI.GetPaymentChannelInfo(ctx, paymentChannel)
+	payChBal := pinfo.Amount.Sub(pinfo.AmountRedeemed)
+
+	if err != nil {
+		return nil, err
+	}
+	if payChBal.LessThan(types.NewAttoFIL(amount.Int)) {
+		return nil, errors.New("not enough funds in payment channel")
+	}
+
+	lanes, ok := r.laneStore[payer]
+	if !ok || len(lanes) == 0 {
+		return nil, errors.New("payment channel has no lanes allocated")
+	}
+
+	// set the allocated amount of the lane
+	lanes[lane] = types.NewAttoFIL(amount.Int)
+
+	v, err := r.createSignedVoucher(ctx, payer, lane, amount)
+	if err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+func (r *RetrievalClientNodeConnector) createSignedVoucher(ctx context.Context, payer address.Address, lane uint64, amount tokenamount.TokenAmount) (*gfm_types.SignedVoucher, error) {
+	height, err := r.getBlockHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	nonce, err := r.actAPI.NextNonce(ctx, payer)
+	if err != nil {
+		return nil, err
+	}
+
+	v := gfm_types.SignedVoucher{
+		TimeLock:       0,   // TODO
+		SecretPreimage: nil, // TODO
+		Extra:          nil, // TODO
+		Lane:           lane,
+		Nonce:          nonce,
+		Amount:         amount,
+		MinCloseHeight: height + 1,
+	}
+
+	var buf bytes.Buffer
+	if err := v.MarshalCBOR(&buf); err != nil {
+		return nil, err
+	}
+
+	signingAddr, err := r.wal.GetDefaultWalletAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	sig, err := r.signer.SignBytes(buf.Bytes(), signingAddr)
+	if err != nil {
+		return nil, err
+	}
+	signature := gfm_types.Signature{
+		Type: gfm_types.KTBLS,
+		Data: sig,
+	}
+	v.Signature = &signature
+	return &v, nil
+}
+
+// handleMessageReceived would be where any subscribers would be notified that the payment channel
+// has been created.
+func (r *RetrievalClientNodeConnector) handleMessageReceived(_ *block.Block, sm *types.SignedMessage, mr *types.MessageReceipt) error {
+	return nil
+}
+
+func (r *RetrievalClientNodeConnector) getBlockHeight() (uint64, error) {
+	head := r.cs.GetHead()
+	ts, err := r.cs.GetTipSet(head)
+	if err != nil {
+		return 0, err
+	}
+	return ts.Height()
+}
+
+func (r *RetrievalClientNodeConnector) getPayerFromChannelID(paymentChannel address.Address) (address.Address, error) {
```

Comment:
Nit: I would rename this to `getPayerForChannel`, and `paymentChannel` to `channelAddr`. The address is the unique identifier in this case (which may or may not be an ID address), but I expect actor addresses to be labeled as addresses, and sequence numbers stored in actor state to be labeled as IDs.

Reply:
That makes sense -- it was confusing to use channelIDs everywhere b/c that's an actual thing right now in go-filecoin, but it's changing with the new spec-actors.